obsplus/events/validate.py

```diff
@@ -100,57 +99,191 @@ def check_origins(event: Event):
 
 
 # register the nullish nslc code replacement
-catalog_validator(replace_null_nlsc_codes)
+validator("obsplus", Event)(replace_null_nlsc_codes)
 
 
-def validate_catalog(events: Union[Catalog, Event],) -> Optional[Union[Catalog, Event]]:
+@validator("obsplus", Event)
+def check_duplicate_picks(event: Event, **kwargs):
+    """
+    Ensure there are no picks with the same phases on the same channels.
+    """
+    # first get dataframe of picks
+    pdf = obsplus.picks_to_df(event)
+    pdf = pdf.loc[pdf.evaluation_status != "rejected"]
+    event_id = str(event.resource_id)
+
+    def dup_picks(phase_hint, on="station"):
+        """ function for checking """
+        df = pdf.loc[pdf.phase_hint == phase_hint]
+        bad = df.loc[df[on].duplicated()][on].tolist()
+        assert len(bad) == 0, (
```

Comment:
As mentioned above, I'm really not a big fan of the use of `assert` statements, but it may be okay in the context of a validation function.

Reply:
I am trying to make the validators much like (py)tests, but that get called at runtime. To maintain parity using assert seems natural (to me). 

Moreover, if you run a group of validators on an object and want to get a report of all the pass/fails it is necessary to either 1) catch all exceptions and assume they are failures and not errors, 2) limit the type of exceptions that validators can raise which indicates an explicit failure rather than some other bug surfacing. We could implement a custom exception, maybe called `ValidationError`, but requiring someone defining their own validators to use it seems a bit onerous to me. Perhaps not though, I am certainly open to more discussion on this. Maybe there is another option I haven't thought of.

Reply:
It makes sense to try to keep parity with testing, and, as I mentioned, maybe it makes sense to use AssertionError in the context of testing. It just runs the risk of getting ignored in some circumstances. Maybe those circumstances can be considered intentional enough that the user should reasonably know what they are doing, though?

I agree that requiring a custom exception, while maybe not terribly onerous, does feel a little clunky.